### ISSUE TITLE
Copying files/ to / before clean-install, fails base image build

### ISSUE THREAD

**Issue by amwat** [kubernetes-sigs/kind#2138]
**What happened**:
```
$ make -C images/base quick
...<redacted>...
#7 0.150 Installing Packages ...
#7 0.231 Err:1 http://archive.ubuntu.com/ubuntu groovy InRelease
#7 0.231   Something wicked happened resolving 'archive.ubuntu.com:80' (-5 - No address associated with hostname)
#7 0.231 Err:2 http://archive.ubuntu.com/ubuntu groovy-updates InRelease
#7 0.231   Something wicked happened resolving 'archive.ubuntu.com:80' (-5 - No address associated with hostname)
#7 0.231 Err:3 http://security.ubuntu.com/ubuntu groovy-security InRelease
#7 0.231   Something wicked happened resolving 'security.ubuntu.com:80' (-5 - No address associated with hostname)
#7 0.232 Err:4 http://archive.ubuntu.com/ubuntu groovy-backports InRelease
#7 0.232   Something wicked happened resolving 'archive.ubuntu.com:80' (-5 - No address associated with hostname)
...<redacted>...
```

**What you expected to happen**:
base image is built successfully

**How to reproduce it (as minimally and precisely as possible)**:
This failed for me on my Google corp workstation, somehow works fine on Mac.

**Anything else we need to know?**:

This is most likely coming from us overwriting some of the `/etc` system files that we copy before running `clean-install`

https://github.com/kubernetes-sigs/kind/blob/426e03350adbe02b65cea8b4ddcfecd8b8119093/images/base/Dockerfile#L59

~wasn't able to pinpoint which one exactly, best guess is https://github.com/kubernetes-sigs/kind/blob/main/images/base/files/etc/sysctl.d/10-network-security.conf affecting the networking somehow?~

Some weird interaction with `/` being the destination ?

**Environment:**

- kind version: (use `kind version`): HEAD
- Kubernetes version: (use `kubectl version`): N/A
- Docker version: (use `docker info`): 20.10.2
- OS (e.g. from `/etc/os-release`):  Debian GNU/Linux

/cc @BenTheElder @aojea

**Comment by BenTheElder** [comment]
> wasn't able to pinpoint which one exactly, best guess is https://github.com/kubernetes-sigs/kind/blob/main/images/base/files/etc/sysctl.d/10-network-security.conf affecting the networking somehow?

If you drop this file does the build work (just delete it from the repo before building)?  Seems like it should be possible to pinpoint it.

Or does it work if you drop all the files? Maybe some race with setting up `/etc/resolv.conf`.

**Comment by aojea** [comment]
Why should fail the DNS resolution with the sysctl file on build? Are we applying the sysctl on build?

**Comment by amwat** [comment]
> If you drop this file does the build work (just delete it from the repo before building)? Seems like it should be possible to pinpoint it.

Unfortunately no, even if I `rm -rf files/etc` the build still fails. There might be a weird interaction with `/` being the destination of `COPY` because separating out the copies does work (as in #2139 )

**Comment by amwat** [comment]
OK, I verified that it is some weird interaction with `/`
just changing `COPY files/ /` to 
```
COPY files/usr/ /usr/
COPY files/etc/ /etc/
```
seems to work.

**Comment by amwat** [comment]
Theres nothing else under `files/` so we can rule out that being a thing.

```
$ ls -al images/base/files/
total 16
drwxr-x--- 4 amwat primarygroup 4096 Mar 17 01:55 .
drwxr-xr-x 3 amwat primarygroup 4096 Mar 17 01:48 ..
drwxr-x--- 6 amwat primarygroup 4096 Mar 17 01:55 etc
drwxr-x--- 3 amwat primarygroup 4096 Mar 17 01:20 usr
```

**Comment by aojea** [comment]
what can I say, it is always dns :upside_down_face: , you copy some files to a container and the thing that fails is dns, it is cursed protocol :rofl:

**Comment by BenTheElder** [comment]
what's fun is it doesn't break on my machine with the same distro. there's small environmental deltas like I have ipv6 enabled on the default docker network.

**Comment by BenTheElder** [comment]
We have a cloud instance of the distro and I can repro with just:

```console
$ tree images/test
images/test
├── Dockerfile
├── files
│   └── etc
└── Makefile

$ cat images/test/Dockerfile
ARG BASE_IMAGE=ubuntu:20.10
FROM $BASE_IMAGE

# copy in static files (configs, scripts)
COPY files/ /

RUN apt-get update && apt-get install openssh
```

I have a feeling this is an issue with the directory tree being setup (note I have no files in `images/test/files/etc`, just the directory), perhaps only happening on some systems due to umask.

**Comment by BenTheElder** [comment]
`umask` reports `0027` on my machine and the cloud instance. The minimal repro also fails on my machine.

this is definitely an issue with the permissions on `/etc` being changed by the `COPY` in some cases.

**Comment by BenTheElder** [comment]
looked at this some more. It's not an issue with docker, the permissions on the files in the clone are different between my local machine and the cloud instance. (755 vs 750)

**Comment by BenTheElder** [comment]
Fun fact: git only stores `0644` or `0755` https://stackoverflow.com/a/40979867, seems git only intends to track executable bit.

So I guess this is a umask issue ultimately. Will need to think a bit about a good fix for this.

**Comment by aojea** [comment]
ADD files.tar

**Comment by BenTheElder** [comment]
I mean we probably don't want to check in a tarball. But we'd need to correctly track permissions on the host somehow

**Comment by aojea** [comment]
IMO the "best" solution is to add ONLY the files we need, that is why  I suggested the tarball.
IIRC that will only change the umask on those files

**Comment by BenTheElder** [comment]
> IMO the "best" solution is to add ONLY the files we need, that is why I suggested the tarball.
> IIRC that will only change the umask on those files

And then we've added some complexity (producing the tarball) while leaving the exact same root problem on the individual file permissions.

What I'm trying to say is that we're not tracking desired file permissions at all, and we need to fix that rather than slapping a bandaid on that makes builds happen to seem to work on @amwat's machine.

This is why we didn't just ship https://github.com/kubernetes-sigs/kind/pull/2139 without root-causing it. This is just sidestepping the issue that we're not tracking intended file modes more generally.

One approach would be top add `COPY` for individual files and take advantage of buildx having `COPY --chmod` to record the desired mode for each file this way (one `COPY --chmod` per file) but that's a bit tedious.

If we just want all of the files / directories to be the same we could e.g. just `chmod -R 0755 ./files/` on the host before building, or `COPY --chmod 0755 ./files/ /`. I'm not sure that we do.

**Comment by BenTheElder** [comment]
In Kubernetes we call `umask 022` in the build scripts before building to avoid the host umask leaking into the build, but here the umask creates differences in the build between machines because of the umask during git operations, not during the build. Whereas in Kubernetes we're applying this to files we generated in part of the build that we will then copy into the image.

**Comment by aojea** [comment]
I'm talking from memory, but I think `tar` allows you to track the permissions and modify them if you want, ... and more stuff,

The complexity here is the same, the only difference is where to put it, in tar or in the Dockerfile, after your explanation, tar is going to be more opaque and makes sense to handle it in the Dockerfile as you said

**Comment by BenTheElder** [comment]
> I'm talking from memory, but I think tar allows you to track the permissions and modify them if you want, ... and more stuff,

Yes but where do we track those?

If we check in a binary tarball we bloat the repo (git is really only designed to track text) and make it difficult to review changes (hard to code review a tarball). So if that's what you mean that seems ... quite less than desirable.

If we don't check in the tarball, where are we recording the permissions we desire? In a script that generates a tarball?
At that point it's just `COPY --chmod` with more steps.

I don't think there's another option involving tar, unless I've missed something?

**Comment by aojea** [comment]
> If we don't check in the tarball, where are we recording the permissions we desire? In a script that generates a tarball?

heh, I got it now, time for a coffee ;)

**Comment by amwat** [comment]
+1 let's avoid worrying about the host settings and be explicit about the file permissions inside the container.
`COPY --chmod` sounds like the best bet and/or leave copy as is and then `chmod` first thing as part of `RUN`

**Comment by BenTheElder** [comment]
For the scripts we do actually currently do:
https://github.com/kubernetes-sigs/kind/blob/426e03350adbe02b65cea8b4ddcfecd8b8119093/images/base/Dockerfile#L91
... which also is missing a new file we added already ...

After sleeping on it I agree, let's do groups of `COPY --chmod` for files similar to the existing PR but with `--chmod` added, and drop the snippet linked above.

**Comment by BenTheElder** [comment]
We already require buildkit / don't support just consuming the dockerfile, and it will be easier to keep the desired permissions in sync on the copy line than down in the `RUN` below and will cache better.

We also should probably not copy directories recursively (see COPY syntax, it's possible).

**Comment by amwat** [comment]
do we know what the file permissions should be? or just everything is 0755 ?
Current looks like
```
$ umask
0027
$ stat -c "%a %n" images/base/files/**
750 images/base/files/
750 images/base/files/etc
750 images/base/files/etc/containerd
640 images/base/files/etc/containerd/config.toml
640 images/base/files/etc/crictl.yaml
750 images/base/files/etc/default
640 images/base/files/etc/default/kubelet
750 images/base/files/etc/sysctl.d
640 images/base/files/etc/sysctl.d/10-network-security.conf
750 images/base/files/etc/systemd
750 images/base/files/etc/systemd/system
640 images/base/files/etc/systemd/system/containerd.service
640 images/base/files/etc/systemd/system/kubelet.service
750 images/base/files/etc/systemd/system/kubelet.service.d
640 images/base/files/etc/systemd/system/kubelet.service.d/10-kubeadm.conf
750 images/base/files/usr
750 images/base/files/usr/local
750 images/base/files/usr/local/bin
750 images/base/files/usr/local/bin/clean-install
750 images/base/files/usr/local/bin/create-kubelet-cgroup-v2
750 images/base/files/usr/local/bin/entrypoint
```

**Comment by amwat** [comment]
I'm guessing we want everything that is `0640` -> `0644` and `0750` -> `0755`

**Comment by BenTheElder** [comment]
Let's start there.

**Comment by BenTheElder** [comment]
We can also check an existing published default node image from v0.10.0. I think the important thing initially is to at least pin specific values so we stop getting arbitrary ones. We can iterate on them.

**Comment by amwat** [comment]
Ack, I think the existing image matches https://github.com/kubernetes-sigs/kind/issues/2138#issuecomment-802244071

```
$ docker run --rm --entrypoint=stat kindest/node:v1.20.2@sha256:8f7ea6e7642c0da54f04a7ee10431549c0257315b3a634f6ef2fecaaedb19bab -c "%a %n" $(find images/base/files -printf '/%P ')
755 /
755 /usr
755 /usr/local
755 /usr/local/bin
755 /usr/local/bin/entrypoint
755 /usr/local/bin/clean-install
755 /usr/local/bin/create-kubelet-cgroup-v2
755 /etc
755 /etc/sysctl.d
644 /etc/sysctl.d/10-network-security.conf
755 /etc/containerd
644 /etc/containerd/config.toml
755 /etc/default
644 /etc/default/kubelet
644 /etc/crictl.yaml
755 /etc/systemd
755 /etc/systemd/system
644 /etc/systemd/system/kubelet.service
755 /etc/systemd/system/kubelet.service.d
644 /etc/systemd/system/kubelet.service.d/10-kubeadm.conf
644 /etc/systemd/system/containerd.service

```
